arcgis-ios-sdk-samples/Route and directions/Navigate route with rerouting/NavigateRouteWithReroutingViewController.swift

```diff
@@ -0,0 +1,364 @@
+    }
+}
+
+// MARK: - AGSRouteTrackerDelegate
+
+extension NavigateRouteWithReroutingViewController: AGSRouteTrackerDelegate {
+    func routeTracker(_ routeTracker: AGSRouteTracker, didGenerateNewVoiceGuidance voiceGuidance: AGSVoiceGuidance) {
+        setSpeakDirection(with: voiceGuidance.text)
+    }
+    
+    func routeTracker(_ routeTracker: AGSRouteTracker, didUpdate trackingStatus: AGSTrackingStatus) {
+        updateTrackingStatusDisplay(routeTracker: routeTracker, status: trackingStatus)
+    }
+    
+    func routeTracker(_ routeTracker: AGSRouteTracker, rerouteDidCompleteWith trackingStatus: AGSTrackingStatus?, error: Error?) {
+        if let error = error {
+            self.presentAlert(error: error)
+        } else {
+            // Get the new directions.
+            directionsList = (trackingStatus?.routeResult.routes.first!.directionManeuvers)!
+        }
+    }
+    
+    func setSpeakDirection(with text: String) {
+        speechSynthesizer.stopSpeaking(at: .word)
+        speechSynthesizer.speak(AVSpeechUtterance(string: text))
+    }
+    
+    func updateTrackingStatusDisplay(routeTracker: AGSRouteTracker, status: AGSTrackingStatus) {
+        var statusText: String
+        switch status.destinationStatus {
+        case .notReached, .approaching:
+            let distanceRemaining = status.routeProgress.remainingDistance.displayText + " " + status.routeProgress.remainingDistance.displayTextUnits.abbreviation
+            let timeRemaining = timeFormatter.string(from: TimeInterval(status.routeProgress.remainingTime * 60))!
+            statusText = """
+            Distance remaining: \(distanceRemaining)
+            Time remaining: \(timeRemaining)
+            """
+            if status.currentManeuverIndex + 1 < directionsList.count {
+                let nextDirection = directionsList[status.currentManeuverIndex + 1].directionText
+                statusText.append("\nNext direction: \(nextDirection)")
+            }
+        case .reached:
+            mapView.locationDisplay.stop()
+            statusText = "Destination reached."
+            routeAheadGraphic.geometry = nil
+        default:
+            statusText = "Off route!"
+        }
+        updateRouteGraphics(remaining: status.routeProgress.remainingGeometry, traversed: status.routeProgress.traversedGeometry)
+        setStatus(message: statusText)
+    }
+    
+    func updateRouteGraphics(remaining: AGSGeometry?, traversed: AGSGeometry? = nil) {
+        routeAheadGraphic.geometry = remaining
+        routeTraveledGraphic.geometry = traversed
+    }
+}
+
+// MARK: - AGSLocationChangeHandlerDelegate
+
+extension NavigateRouteWithReroutingViewController: AGSLocationChangeHandlerDelegate {
+    func locationDataSource(_ locationDataSource: AGSLocationDataSource, locationDidChange location: AGSLocation) {
+        // Update the tracker location with the new location from the simulated data source.
+        routeTracker?.trackLocation(location)
```

Comment:
This might be an excessive request, but consider handling the error here. ref to common-samples/issues/2289. (Other 2 navigate route and AR samples yet to be fixed)
We eventually want to ensure users being aware of the licensing requirement by the routeTracker.
The simplest way to handle is to nilify the delegate of routeTracker whenever an error occurs.

Reply:
I gave it a try, please let me know what you think!